NDTensors/src/dense/fill.jl

```diff
@@ -21,19 +21,26 @@ function generic_randn(StoreT::Type{<:Dense}, dim::Integer=0)
   return generic_randn(default_storagetype(), dim)
 end
 
-function generic_zeros(
-  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
-) where {DataT<:AbstractArray,ElT}
-  @assert ElT == eltype(DataT)
-  data = generic_zeros(DataT, dim)
-  StoreT = set_datatype(StoreT, typeof(data))
-  return StoreT(data)
+function generic_zeros(StoreT::Type{<:Dense}, dim::Integer)
```

Comment:
I think this should become a generic definition for `generic_zeros` for any `Type{<:AbstractArray}`.

Reply:
Right now you have organized the code like this:
```julia
# Implementation, catches if `ndims(arraytype) != length(dims)`.
function generic_zeros(arraytype::Type{<:AbstractArray}, dims...)

# Forwards to `generic_zeros(::Type{<:AbstractArray}, ::Tuple)`
function generic_zeros(::Type{<:AbstractArray}, dims...)

# Forwards to `generic_zeros(::Type{<:Dense}, ::Tuple)`
function generic_zeros(StoreT::Type{<:Dense}, dim::Integer)

# Implementation
function generic_zeros(::Type{<:Dense}, dims::Tuple{Integer})

# Catches errors
function generic_randn(::Type{<:Dense}, ::Tuple{Integer,Integer,Vararg{Integer}})
```
You can see that it is inconsistent which signature is responsible for the implementation and which signature is responsible for catching dimension errors in the generic case and in the `Dense` case. I'm suggesting organizing it like this:
```julia
# Implementation, catches if `ndims(arraytype) != length(dims)`.
function generic_zeros(arraytype::Type{<:AbstractArray}, dims::Tuple)

# Forwards to `generic_zeros(::Type{<:AbstractArray}, ::Tuple)`
function generic_zeros(::Type{<:AbstractArray}, dims...)

# Implementation
function generic_zeros(::Type{<:Dense}, dims::Integer)
```
If my logic is correct, then `generic_zeros(Dense, (2,))` will forward to `generic_zeros(Dense, 2)` through the `generic_zeros(::Type{<:AbstractArray}, ::Tuple)` function, and calls like `generic_zeros(Dense, (2, 2))` or `generic_zeros(Dense, 2, 2)` will end up in the generic `generic_zeros(arraytype::Type{<:AbstractArray}, dims::Tuple)` version and will throw an error because of a dimension mismatch.

Reply:
Ahh yes that is something I didn't check. I updated the code and now it no longer throws an error! One question is that I found in this new structure is that 
`generic_zeros(Dense, (2,3)) = Error: Can't make a multidimensional `Dense` object.` which is what we expect. But when I do this `generic_zeros(Dense, 2,3)` it calls the `AbstractArray` version which ultimately fails but is unexpected behavior. To fix this I have these two catches.
```
function generic_zeros(::Type{<:Dense}, ::Integer, ::Vararg{Integer})
  return error("Can't make a multidimensional `Dense` object.")
end

function generic_zeros(::Type{<:Dense}, ::Tuple{Integer,Integer,Vararg{Integer}})
  return error("Can't make a multidimensional `Dense` object.")
end
```
Does this seem reasonable?

Reply:
The way I would design it would be like this:
```julia
# Generic `AbstractArray` implementation, also catches if `ndims(arraytype) != length(dims)`.
function generic_zeros(arraytype::Type{<:AbstractArray}, dims::Integer...) ... end

# Forwards to `generic_zeros(::Type{<:AbstractArray}, ::Integer...)`
function generic_zeros(::Type{<:AbstractArray}, dims::Tuple{Vararg{Integer}}) ... end

# `Dense` implementation.
function generic_zeros(::Type{<:Dense}, dims::Integer) ... end
```
Does that work and have all of the behavior we want?

Reply:
@kmp5VT what about this final design I suggested? I think the code you wrote is still a bit more complicated than it needs to be. Also I think we should make `generic_zeros(arraytype::Type{<:AbstractArray}, dims::Integer...)` in inner implementation function and forward to that from `generic_zeros(arraytype::Type{<:AbstractArray}, dims::Tuple{Vararg{Integer}})`, that is more standard for how the analgous Base Julia functions are defined.

Reply:
Hi sorry I missed this comment on my last go through. The issue with the design as you have written it is that codes like
```
generic_zeros(Dense, 2,3)
generic_zeros(Dense, (2,3))
```
call the `AbstractArray` version which is incorrect behavior.  I added the extra code to catch these cases and throw an error.

I swapped the function, though this requires rewrititng the dense function like this 
```
function generic_zeros(StoreT::Type{<:Dense}, dims::Tuple{Integer})
```

Reply:
Right, but then wouldn't it catch that `ndims(Dense) != length((2, 3))`, since `ndims(Dense) == 1`?

Reply:
No I am not using `ndims` I am using `type_parameter(Dense, ndims) = ERROR: Not implemented` because Dense doesn't have an `ndims` parameter

Reply:
I see, but that seems like more an issue of how you've implemented `generic_zeros(arraytype::Type{<:AbstractArray}, dims...)`.

Reply:
I.e. maybe it shouldn't be calling `type_parameter(Dense, ndims)`, and instead should call `ndims(Dense)`.

Reply:
Hopefully you understand my motivation for harping on this point, if we make the generic version work better and more general it makes it easier to define specialized versions for new array types, just like our over all goal with the GPU code of making better generic code to lower the barrier for defining new GPU backends.

Reply:
Using `ndims(arraytype_specified)` doesn't fix the issue because the first thing that happens in both of these `generic_x(::Type{<:AbstractArray})` functions is `arraytype_specified = specify_type_parameter(unwrap_array_type(arraytype), ndims, length(dims)` and `Dense` does not have a position for `ndims` defined. And though we have a `set_ndims` function defined, we don't have a `specify_ndims` function. 

I totally understand the goal and appreciate your perspective! It helps with seeing holes in my logic and finding more robust solutions.

Reply:
Right, that just seems to be another indication that the logic of `generic_x(::Type{<:AbstractArray}, ...)` isn't quite correct or general enough and should be rethought, since it is a very reasonable case to input a vector or matrix type that has a hard-coded `ndims` so doesn't have an `ndims` type parameter.